dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/DependencyConflict.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.cloud.tools.opensource.dependencies.Artifacts;
+import com.google.cloud.tools.opensource.dependencies.DependencyPath;
+import java.util.Objects;
+import org.eclipse.aether.artifact.Artifact;
+
+/**
+ * Diamond dependency conflict caused a {@link LinkageProblem} where the {@link LinkageProblem}'s
+ * invalid reference points to the symbol in {@code pathToSelectedArtifact.getLeaf()} but a valid
+ * symbol is in {@code pathToUnselectedArtifact.getLeaf()}.
+ */
+class DependencyConflict extends LinkageProblemCause {
+  DependencyPath pathToUnselectedArtifact;
+  DependencyPath pathToSelectedArtifact;
+
+  DependencyConflict(
+      DependencyPath pathToSelectedArtifact, DependencyPath pathToUnselectedArtifact) {
+    this.pathToUnselectedArtifact = checkNotNull(pathToUnselectedArtifact);
+    this.pathToSelectedArtifact = checkNotNull(pathToSelectedArtifact);
+  }
+
+  public DependencyPath getPathToUnselectedArtifact() {
+    return pathToUnselectedArtifact;
+  }
+
+  public DependencyPath getPathToSelectedArtifact() {
+    return pathToSelectedArtifact;
+  }
+
+  @Override
+  public String toString() {
+    Artifact selected = pathToSelectedArtifact.getLeaf();
+    Artifact unselected = pathToUnselectedArtifact.getLeaf();
+
+    return "Dependency conflict: "
+        + Artifacts.toCoordinates(selected)
+        + " was in the class path "
+        + "but version '"
+        + unselected.getVersion()
+        + "' has a valid symbol.\n"
+        + "  selected: "
+        + pathToSelectedArtifact
+        + "\n  unselected: "
+        + pathToUnselectedArtifact;
```

Comment:
Example output:
```
(com.google.guava:guava:20.0) com.google.common.base.Verify's method verify(boolean arg1, String arg2, Object arg3) is not found;
  referenced by 3 class files
    io.grpc.internal.ServiceConfigInterceptor (io.grpc:grpc-core:1.17.1)
    io.grpc.internal.JndiResourceResolverFactory (io.grpc:grpc-core:1.17.1)
    io.grpc.internal.DnsNameResolver (io.grpc:grpc-core:1.17.1)
  Cause:
    Dependency conflict: com.google.guava:guava:20.0 was in the class path but version '26.0-android' has a valid symbol.
      selected: com.google.api-client:google-api-client:1.27.0 (compile) / com.google.guava:guava:20.0 (compile)
      unselected: io.grpc:grpc-core:1.17.1 (compile) / com.google.guava:guava:26.0-android (compile)
```

Note that when there's an invalid reference between ArtifactX and ArtifactY, it's not always true that we should upgrade ArtifactY. Users may have to upgrade or downgrade Artifact X or Artifact Y.